src/server/game/Entities/Player/Player.cpp

```diff
@@ -1937,7 +1937,10 @@ bool Player::TeleportTo(uint32 mapid, float x, float y, float z, float orientati
         {
             Position oldPos = GetPosition();
             if (HasUnitMovementFlag(MOVEMENTFLAG_HOVER))
-                z += GetFloatValue(UNIT_FIELD_HOVERHEIGHT);
+            {
+                float ground = GetMap()->GetHeight(GetPhaseMask(), x, y, z);
```

Comment:
Yes, a unit can be TPed anywhere, including in the air. However, for a unit without Hover & DisableGravity flag, as soon as such unit is put at a new position in the air, the client will apply a force (gravity) on that unit until the unit reaches the ground.

It works more or less the same way for hovering units: As long as a unit is not at z=ground_height + hover_height, the client will apply a force on that unit to make the unit move to that stable position.

Ok, that's great, but where am I going with this? Well, there is something missing in the core. If the client applies a force on units controlled by the server, the server needs to do the same. Otherwise, the unit will move in the client but not in the server. And a desynchronization of the unit's position as seen by the server and by the client will occur.

In order to circonvent this problem (and until such time that gravity on falling creature  is implemented), this PR's approach is to avoid putting creatures in "unstable" positions. And as long as creatures are in "stable" positions (= on the ground for units subjected to gravity, = ground_height + hover_height for hovering units), all is well.

I'm not saying this approach is perfect. I will see what I can do next.